### PR TITLE
[api] Also deprecate constructors of Decl_kinds.

### PR DESCRIPTION
**Pull request by ejgallego**
Unfortunately OCaml doesn't deprecate the constructors of a type when
the type alias is deprecated.

In this case it means that we don't get rid of the kernel dependency
unless we deprecate the constructors too.

### CODE REVIEW

**Review by SkySkimmer — APPROVED** [review]

File: API/API.mli [hunk 1]
```diff
@@ -2160,10 +2160,14 @@ module Decl_kinds :
...
+
   type recursivity_kind = Declarations.recursivity_kind =
-    | Finite
-    | CoFinite
-    | BiFinite
+    | Finite (** = inductive *)
+        [@ocaml.deprecated "Please use [Declarations.Finite"]
+    | CoFinite (** = coinductive *)
+        [@ocaml.deprecated "Please use [Declarations.CoFinite"]
+    | BiFinite (** = non-recursive, like in "Record" definitions *)
```
Comment (SkySkimmer) [inline API/API.mli]: These comments should added to the non deprecated type.

Comment (ejgallego) [reply API/API.mli]: Updated. Anyways I hope the kernel part of API is going away soon.

**Comment by maximedenes** [comment]
@ppedrot is that one ok with you?

**Review by ppedrot — APPROVED** [review]
Good to go.

**Comment by maximedenes** [comment]
This conflicts with https://github.com/coq/coq/pull/6318, which I will schedule first because it has been delayed many times already, and you agreed it was prioritary.

**Comment by ejgallego** [comment]
Sure, all this deprecation commits are very easy to rebase, feel free to delay them.

**Comment by ejgallego** [comment]
[when a conflict happens, I basically rebase, then re-launch my "automatic" warning fixer]

**Comment by SkySkimmer** [comment]
What does the warning fixer do?

**Comment by ejgallego** [comment]
> What does the warning fixer do?

It's a hellish mix of emacs `(compile)` + `(merlin-next-warning)` + custom stuff such as replacing a name or whatever is needed in each case.

**Pull request merged**